### PR TITLE
MBS-7281: Load learningmap via webservice

### PR DESCRIPTION
**Pull request by PhMemmel**
This makes the learningmap being loaded by a webservice.

Also I implemented a reactive component watching the current state of depending activities: If completion state changes or activity is being renamed, the learningmap (and only the map) updates itself without reloading the whole page.

### CODE REVIEW

**Comment by sh-csg** [comment]
Thanks a lot for this great improvement! It is an excellent alternative to my iframe suggestion!
Do you see any way to keep compatibility with clients without Javascript?

**Comment by PhMemmel** [comment]
Thanks. I tried to think of a way supporting browsers without JS, but I do not think, there is one.

Using noscript would mean we have to call the render function always before the learningmap is being sent to the client browser which is exactly what we want to fix with this. Aside from that I do not see how the PHP backend would be able to find out beforehand if the client supports JS or not. However, looking at how extensively moodle core uses JS and having the year 2023 I think it would be safe to drop support for browser that do not support JS. Or could you think of an edge case where this would be a thing?

File: amd/src/renderer.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            {
+                methodname: 'mod_learningmap_get_learningmap',
+                args: {
+                    'cmId': cmId
+                }
+            }
+        ]);
+
+    promises[0].then(data => {
+        const targetDiv = document.getElementById(
```
Comment (sh-csg) [inline amd/src/renderer.js:60]: Fits in one line

File: amd/src/manual-completion-watch.js [hunk 2]
```diff
@@ -1,16 +1,28 @@
...
+
+/**
+ * This module only servers backwards compatibility for moodle versions 3.9 and 3.11.
+ * As soon as the support is being dropped, this can be removed as well as the call in lib.php::learningmap_cm_info_view
+ *
+ * @param {array} coursemodules course modules the learningmap depends on
+ */
 export const init = (coursemodules) => {
     coursemodules.forEach((c) => {
-        let selector = '[data-action="toggle-manual-completion"][data-cmid="' + c + '"]';
```
Comment (sh-csg) [inline amd/src/manual-completion-watch.js:3]: These lines are necessary to keep compatibility with moodle 3.11 as it already uses the new completion buttons.

File: templates/rendercontainer.mustache [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+<div id="learningmap-render-container-{{cmId}}" class="learningmap-render-container">
+  {{> mod_learningmap/loadingplaceholder}}
+</div>
+{{#js}}
+  require(['mod_learningmap/renderer'], function(renderer) {
+    renderer.init({{cmId}});
+  });
+  {{#isIntro}}
+      require(['mod_learningmap/initliveupdater'], function(initliveupdater) {
+      initliveupdater.init({{cmId}});
```
Comment (sh-csg) [inline templates/rendercontainer.mustache:47]: Indentation

**Review by sh-csg — COMMENTED** [review]
Looks really good for now. Tested also with 3.9 / 3.11.
Blocker for now: Breaks installation in 3.9 and manual-completion in 3.11
Apart from that only minor problems.

**Comment by sh-csg** [comment]
> Thanks. I tried to think of a way supporting browsers without JS, but I do not think, there is one.
> 
> Using noscript would mean we have to call the render function always before the learningmap is being sent to the client browser which is exactly what we want to fix with this. Aside from that I do not see how the PHP backend would be able to find out beforehand if the client supports JS or not. However, looking at how extensively moodle core uses JS and having the year 2023 I think it would be safe to drop support for browser that do not support JS. Or could you think of an edge case where this would be a thing?

No real idea how to solve this - just for thinking about it. Might not be a relevant case (it maybe just came into my mind because I was handling some Behat tests failing when JS was turned off).

File: amd/src/manual-completion-watch.js (hunk 2 above)
Comment (PhMemmel) [reply amd/src/manual-completion-watch.js:3]: Sorry, thought this is 4.0 only. Reverted.

File: db/services.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ * @author     Philipp Memmel
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+defined('MOODLE_INTERNAL') || die;
+
+$functions = [
+    'mod_learningmap_get_learningmap' => [
+        'classname'   => 'mod_learningmap\external\get_learningmap',
+        'description' => 'Retrieves the learningmap',
```
[the remark this answers is not in the document]
Comment (PhMemmel) [reply db/services.php]: Yes, this is the default anyway, so I left it out, but apparently not yet on 3.9. Thanks for testing.

File: db/services.php [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        'type'        => 'read',
+        'ajax' => true,
+        'capabilities' => '',
```
[the remark this answers is not in the document]
Comment (PhMemmel) [reply db/services.php]: I left this out on purpose, because I'm doing the capability check anyway in the external function itself. This would allow me to deal with the issue of missing capability directly there and handle it individually. But of course one could double check it here.

File: db/services.php [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    ],
+    'mod_learningmap_get_dependingmodules' => [
+        'classname'   => 'mod_learningmap\external\get_dependingmodules',
+        'description' => 'Retrieves the course modules a given learningmap depends on',
+        'type'        => 'read',
+        'ajax' => true,
+        'capabilities' => '',
```
[the remark this answers is not in the document]
Comment (PhMemmel) [reply db/services.php]: see above

File: lib.php [hunk 7]
```diff
@@ -226,10 +226,21 @@ function learningmap_cm_info_view(cm_info $cm) : void {
                 $groupdropdown
             );
         }
-        $cm->set_content($groupdropdown . learningmap_get_learningmap($cm), true);
+
+        $isintro = true;
```
[the remark this answers is not in the document]
Comment (PhMemmel) [reply lib.php]: Yes, that is not ideal. I decided to change it completely to be able to just define if the liveupdater should be loaded or not.
